protocol/kona-node-arch.md

```diff
@@ -0,0 +1,291 @@
+│L2 Sequencer│
+│            ├────────────────────┐
+│   Gossip   │                    │    ┌────────────┐
+└────────────┘                    │    │            │
+                                  ├──► │     ???    │──►  < L2 Chain >
+┌────────────┐    ┌────────────┐  │    │            │
+│     DA     │    │            │  │    └────────────┘
+│            ├──► │ Derivation ├──┘
+│   Watcher  │    │            │
+└────────────┘    └────────────┘
+```
+
+From these sources, the rollup node imports "unsafe" blocks from the L2 sequencer
+and safe blocks from the L2 derivation pipeline. Both unsafe and safe blocks
+are imported into the L2 execution layer via the Engine API.
+
+```
+┌────────────┐
+│L2 Sequencer│
+│            ├────────────────────┐
+│   Gossip   │                    │    ┌────────────┐
+└────────────┘                    │    │            │
+                                  ├──► │ Engine API │──►  < L2 Chain >
+┌────────────┐    ┌────────────┐  │    │            │
+│     DA     │    │            │  │    └────────────┘
+│            ├──► │ Derivation ├──┘
+│   Watcher  │    │            │
+└────────────┘    └────────────┘
+```
+
+The [Holocene Hardfork][holocene] introduced steady block derivation.
+This change allows a payload to be replaced with a deposits-only block
+to allow the L2 chain to progress even if an invalid payload is derived.
+
+Steady block derivation has significant architectural impacts for the
+rollup node. With Holocene, the derivation component needs to be updated
+if a payload is replaced with a deposits-only block. This requires the
+Engine API component to send a message to the derivation component,
+introducing an edge in the modular component graph.
+
+[holocene]: https://specs.optimism.io/protocol/holocene/derivation.html
+
+```
+┌────────────┐
+│L2 Sequencer│
+│            ├───┐
+│   Gossip   │   │   ┌────────────┐   ┌────────────┐
+└────────────┘   │   │            │   │            │
+                 ├──►│ Derivation │──►│ Engine API │──►  < L2 Chain >
+┌────────────┐   │   │            │   │            │
+│     DA     │   │   └────────────┘   └──────┬─────┘
+│            ├───┘          ▲                │
+│   Watcher  │              └────────────────┘
+└────────────┘
+```
+
+### A Modular Architecture
+
+The components described above are all critical to the rollup node.
+Whether these components are part of the architecture is not up for discussion.
+What this document addresses is how the components communicate. Namely, do
+components share memory or use an actor=based architecture with message-passing.
+The remainder of this section will provide reasoning to choose the latter.
+
+In an actor-based system, each component is constructed as an "Actor". There's
+a Derivation Actor, DA Watcher Actor, P2P Actor (L2 sequencer gossip),
+and Engine Actor.
+
+Instead of having some top-level object "own" components, actors are spawned
+as threads, and communication between actors happens through channels with
```

Comment:
are these channels synchronous? If not, what's the queueing policy and how do events/ message receipts handled?

Reply:
Good question. The answer is it really depends on the actors and the relationships between those actors. Since the channels don't have to be uniform, I haven't specified anything here. Most likely, it'll change as we go along.